### PR TITLE
Destructuring (close #642)

### LINKED ISSUE #642: The wrong processing of destructuring
If `location` declared inside function or passed as function parameter we should not process it.
It needs for fix https://github.com/DevExpress/testcafe-hammerhead/issues/640#issuecomment-228679853

**Update!** The origin issue was transfered to https://github.com/DevExpress/testcafe-hammerhead/issues/2283

### PR DESCRIPTION
**Pull request by LavrovArtem**

### CODE REVIEW

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

File: src/processing/script/transformers/temp-variables.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import { Expression } from 'estree';
+import { Syntax } from 'esotope-hammerhead';
+
+export default class TempVariables {
+    private static _tempVariableId = 0;
```
Comment (miherlosev) [inline src/processing/script/transformers/temp-variables.ts:5]: `private static _counter = 0;`

File: src/processing/script/transformers/temp-variables.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    private _variables: string[] = [];
+
+    static resetId () {
```
Comment (miherlosev) [inline src/processing/script/transformers/temp-variables.ts:8]: `static resetCounter () `

File: src/processing/script/transformers/temp-variables.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        TempVariables._tempVariableId = 0;
+    }
+
+    static generate(baseName?: string, key?: Expression, index?: number): string {
```
Comment (miherlosev) [inline src/processing/script/transformers/temp-variables.ts:12]: `static generateName(`

File: src/processing/script/transformers/temp-variables.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import { Expression } from 'estree';
+import { Syntax } from 'esotope-hammerhead';
+
+export default class TempVariables {
+    private static _tempVariableId = 0;
+    private _variables: string[] = [];
```
Comment (miherlosev) [inline src/processing/script/transformers/temp-variables.ts:6]: `private _list: string[] = [];`

File: src/processing/script/index.ts [hunk 5]
```diff
@@ -140,14 +140,14 @@ function applyChanges (script: string, changes: CodeChange[], isObject: boolean)
     if (!changes.length)
         return script;
 
-    changes.sort((a, b) => a.start - b.start);
+    changes.sort((a, b) => (a.start - b.start) || (a.end - b.end));
```
Comment (miherlosev) [inline src/processing/script/index.ts:143]: What does it mean?

File: src/processing/script/transformers/index.ts [hunk 6]
```diff
@@ -17,26 +18,28 @@ import windowPostMessageGetTransformer from './window-post-message-get';
...
 import dynamicImportTransformer from './dynamic-import';
+import declarationDestructuring from './declaration-destructuring';
+import assignmentDestructuring from './assignment-destructuring';
 
 export interface Transformer<C extends Node> {
     nodeReplacementRequireTransform: boolean;
     nodeTypes: C['type'];
     condition: (node: C, parent?: Node) => boolean;
-    run: <P extends Node>(node: C, parent?: P, key?: keyof P) => Node|null;
+    run: <P extends Node>(node: C, parent?: P, key?: keyof P, tempVars?: TempVariables) => Node|null;
```
Comment (miherlosev) [inline src/processing/script/transformers/index.ts:35]: Do we will remove `__set$temp` and fix case for the nested `for-in` in separate PRs?

File: test/server/script-processor-test.js [hunk 7]
```diff
@@ -804,6 +804,118 @@ describe('Script processor', () => {
         ]);
     });
 
+    describe('Destructuring', () => {
+        it('Should process object pattern declaration', () => {
```
Comment (miherlosev) [inline test/server/script-processor-test.js:808]: I guess we can remove the `Should process` phrase from all test names.

File: src/processing/script/index.ts (hunk 5 above)
Comment (LavrovArtem) [reply src/processing/script/index.ts:143]: This change needs for the following array:
```js
[
    { start: 0, end: 20, code: 'a=temp.a, b=temp.b' },
    { start: 0, end: 0, code: 'var temp=obj' }
] // changes
```

File: src/processing/script/transformers/index.ts (hunk 6 above)
Comment (LavrovArtem) [reply src/processing/script/transformers/index.ts:35]: Yes, I will do it in separate PRs.

File: src/processing/script/destructuring.ts [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    return createSimpleCallExpression(restArrayIdentifier, [array, createSimpleLiteral(startIndex)]);
+}
+
+function processObjectPattern (pattern: ObjectPattern, value: Expression, build: NodeBuilder, baseTempName: string) {
+    if (!baseTempName)
+        baseTempName = TempVariables.generateName(baseTempName);
+
+    const properties     = pattern.properties;
+    // @ts-ignore
+    const hasRest        = properties[properties.length - 1].type === Syntax.RestElement;
```
Comment (AndreyBelym) [inline src/processing/script/destructuring.ts:53]: Is it guaranteed that length is always greater than 1? Empty destructuring is an edge case, but it is a valid expression:
```js
console.log({}={a:5})
```

**Pull request merged**